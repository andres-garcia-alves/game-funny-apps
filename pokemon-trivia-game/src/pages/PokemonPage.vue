<template>
    <div>
        <div v-if="pokemon">
            <h1>¿Quién es este pokemón?</h1>

            <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon" />
            <PokemonOptions :pokemons="pokemonArray" @selection="checkAnswer($event)" />

            <template v-if="showAnswer">
                <h2 class="fade-in">{{ message }}</h2>
                <button @click="newGame">Nuevo juego</button>
            </template>
        </div>       
        <h2 v-else>espere por favor...</h2>
    </div>
</template>

<script>
import PokemonPicture from '@/components/PokemonPicture.vue'
import PokemonOptions from '@/components/PokemonOptions.vue'
import getPokemonOptions from '@/helpers/getPokemonOptions'

export default {
    name: 'PokemonPage',
    components: {
        PokemonPicture,
        PokemonOptions
    },
    data() {
        return {
            pokemonArray: [],
            pokemon: null,
            showPokemon: false,
            showAnswer: false,
            message: ''
        }
    },
    methods: {
        async mixPokemonArray() {
            this.pokemonArray = await getPokemonOptions()
            const idx = Math.floor(Math.random() * 4)

            this.pokemon = this.pokemonArray[idx]
        },
        checkAnswer(selectedId) {
            this.showPokemon = true
            this.showAnswer = true
            this.message = (selectedId === this.pokemon.id) ?
                `Correcto, ${ this.pokemon.name }.` :
                `Ups, en realidad es: ${ this.pokemon.name }.`
        },
        async newGame() {
            this.pokemonArray = []
            this.pokemon = null
            this.showPokemon = false
            this.showAnswer = false
            this.message = ''

            await this.mixPokemonArray()
        }
    },
    mounted() {
        this.mixPokemonArray()
    }
}
</script>

<style scoped>
button {
    background-color: black;
    border-radius: 5px;
    border: 1px solid rgba(0, 0, 0, 0.2);
    color: white;
    cursor: pointer;
    margin-bottom: 20px;
    padding: 5px;
    width: 250px;
}
</style>
