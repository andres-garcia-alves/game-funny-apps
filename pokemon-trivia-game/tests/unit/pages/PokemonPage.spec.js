import { shallowMount, mount  } from "@vue/test-utils";

import pokemonsMock from "../mocks/pokemons.mock";
import PokemonPage from "@/pages/PokemonPage";

describe("pages/PokemonPage.vue", () => {
    
    test("debe coincidir con el snapshot", () => {
        const wrapper = shallowMount(PokemonPage)
        expect(wrapper.html()).toMatchSnapshot()
    })

    test("debe coincidir con el snapshot luego de cargado los pokemons", () => {
        const wrapper = mount(PokemonPage, {
            data() {
                return { pokemonArray: pokemonsMock, pokemon: pokemonsMock[0], showPokemon: false, showAnswer: false, message: '' }
            }
        })

        expect(wrapper.html()).toMatchSnapshot()
    })

    test("debe de llamarse mixPokemonArray() al montarse el componente", () => {
        const mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray')
        const wrapper = shallowMount(PokemonPage)

        expect(mixPokemonArraySpy).toHaveBeenCalledTimes(1)
    })

    test("debe de mostrar los componentes PokemonPicture y PokemonOptions", () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return { pokemonArray: pokemonsMock, pokemon: pokemonsMock[0], showPokemon: false, showAnswer: false, message: '' }
            }
        })

        const pictureStub = wrapper.find('pokemon-picture-stub')
        const optionsStub = wrapper.find('pokemon-options-stub')

        expect(pictureStub.exists()).toBe(true)
        expect(optionsStub.exists()).toBe(true)

        expect(pictureStub.attributes('pokemonid')).toBe('5')
        expect(optionsStub.attributes('pokemons')).toBeTruthy()
    })

    test("debe validarse checkAnswer()", async () => {
        // arrange
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return { pokemonArray: pokemonsMock, pokemon: pokemonsMock[0], showPokemon: false, showAnswer: false, message: '' }
            }
        })

        // act (con 'await' para esperar que Vue aplique los cambios en el DOM virtual)
        await wrapper.vm.checkAnswer(5)

        // asserts
        const h2 = wrapper.find('h2')
        expect(h2.exists()).toBe(true)
        expect(h2.text()).toBe(`Correcto, ${ pokemonsMock[0].name }.`)
        expect(wrapper.vm.showPokemon).toBe(true)

        // act (con 'await' para esperar que vue aplique los cambios en el DOM virtual)
        await wrapper.vm.checkAnswer(10)

        // asserts (esta vez sin usar el 'h2', sino con la propiedad 'message')
        expect(wrapper.vm.message).toBe(`Ups, en realidad es: ${ pokemonsMock[0].name }.`)
    })
})
