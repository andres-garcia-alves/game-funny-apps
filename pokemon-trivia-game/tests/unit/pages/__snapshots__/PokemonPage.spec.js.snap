// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/PokemonPage.vue debe coincidir con el snapshot 1`] = `
<div>
  <h2>espere por favor...</h2>
</div>
`;

exports[`pages/PokemonPage.vue debe coincidir con el snapshot luego de cargado los pokemons 1`] = `
<div>
  <div>
    <h1>¿Quién es este pokemón?</h1>
    <div class="pokemon-container"><img src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/5.svg" class="hidden-pokemon" alt="pokemon"></div>
    <div class="options-container">
      <ul>
        <li>pikachu</li>
        <li>charmander</li>
        <li>venusaur</li>
        <li>mew</li>
      </ul>
    </div>
    <!--v-if-->
  </div>
</div>
`;
